### PR TITLE
standalone implementation of ccalendar

### PR DESCRIPTION
**Pull request by jbrockmendel**
For the moment only updated cimports in `fields`.

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
what PR was this originally implemented in?

File: pandas/_libs/tslibs/ccalendar.pyx [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    -----
+    Assumes that the arguments are valid.  Passing a month not between 1 and 12
+    risks a segfault.
+    """
+    return days_per_month_array[12 * is_leapyear(year) + month - 1]
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef monthrange(int64_t year, Py_ssize_t month):
```
Comment (jreback) [inline pandas/_libs/tslibs/ccalendar.pyx]: why do we need this as a function that returns a tuple. its it simply better to split into 2, e.g. monthday and dayofweek? its 2 functions, but then these become all nogil

File: pandas/_libs/tslibs/ccalendar.pyx [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    Notes
+    -----
+    Assumes that y, m, d, represents a valid date.
+    """
+    cdef:
+        int day
+
+    y -= m < 3
+    day = (y + y / 4 - y / 100 + y / 400 + sakamoto_arr[m - 1] + d) % 7
+    # convert to python day
```
Comment (jreback) [inline pandas/_libs/tslibs/ccalendar.pyx]: add a link to python dayofweek docs (e.g. its sunday 0, -6)

File: pandas/_libs/tslibs/ccalendar.pyx [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    return (day + 6) % 7
+
+
```
Comment (jreback) [inline pandas/_libs/tslibs/ccalendar.pyx]: on get_week_of_year you have boundscheck/wraparound but not here, do we need in either?

File: pandas/_libs/tslibs/ccalendar.pxd [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
+# cython: profile=False
```
Comment (jreback) [inline pandas/_libs/tslibs/ccalendar.pxd]: is this *only* used in fields? no-where else?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/ccalendar.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/ccalendar.pyx]: Those two separate functions also exist here.  There's not much reason to retain monthrange, you're right.

File: pandas/_libs/tslibs/ccalendar.pxd (hunk 4 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/ccalendar.pxd]: Only implemented in fields.  Haven't updated others yet.  A couple places in offsets, a few more fastpaths in Timestamp.

File: pandas/_libs/tslibs/ccalendar.pyx (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/ccalendar.pyx]: Is this comment about `is_leapyear`?  If so, is_leapyear doesn't do any array lookups, so those decorators are irrelevant.

Comment (jreback) [reply pandas/_libs/tslibs/ccalendar.pyx]: yea

**Comment by jbrockmendel** [comment]
> what PR was this originally implemented in?

#18489, but that was reverted as out-of-scope

**Comment by jbrockmendel** [comment]
Just pushed a commit that uses `ccalendar` functions to implement fastpaths for some of the `Timestamp` properties (ones not covered by #18539)

master
```
In [2]: ts = pd.Timestamp.now()
In [3]: %timeit ts.is_month_end
The slowest run took 34.08 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 3: 26.8 µs per loop
```

PR 
```
In [2]: ts = pd.Timestamp.now()
In [3]: %timeit ts.is_month_end
The slowest run took 27.70 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 1.19 µs per loop
```

**Review by jreback — APPROVED** [review]

File: pandas/_libs/tslibs/ccalendar.pyx (hunk 1 above)
Comment (jreback) [reply pandas/_libs/tslibs/ccalendar.pyx]: are you changing this in this PR?

File: pandas/_libs/tslibs/ccalendar.pyx [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    Returns
+    -------
+    is_leap : bool
+    """
+    return ((year & 0x3) == 0 and  # year % 4 == 0
+            ((year % 100) != 0 or (year % 400) == 0))
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
```
Comment (jreback) [inline pandas/_libs/tslibs/ccalendar.pyx]: you changing the boundscheck/wraparound in this PR?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/ccalendar.pyx (hunk 5 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/ccalendar.pyx]: Hadn't planned on it.  Are they harmful?

**Comment by jbrockmendel** [comment]
> are you changing [monthrange] in this PR?

Hadn't planned on it.  Even if it shouldn't be used, ATM it still is used in a few places.  Besides I wrote a really nice docstring.

**Comment by jreback** [comment]
> Hadn't planned on it. Even if it shouldn't be used, ATM it still is used in a few places. Besides I wrote a really nice docstring.

what changed? we had this discussion about why returning a tuple is a bad idea

File: pandas/_libs/tslibs/ccalendar.pyx (hunk 5 above)
Comment (jreback) [reply pandas/_libs/tslibs/ccalendar.pyx]: prob not harmful but extra code and confusing. pls change.

File: pandas/_libs/tslibs/ccalendar.pyx [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        raise ValueError("bad month number 0; must be 1-12")
+
+    days = get_days_in_month(year, month)
+    return (dayofweek(year, month, 1), days)
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.cdivision
+cdef int dayofweek(int y, int m, int d) nogil:
```
Comment (jorisvandenbossche) [inline pandas/_libs/tslibs/ccalendar.pyx]: It seems you are duplicating things from `np_datetime.c` ?

**Comment by jbrockmendel** [comment]
> what changed? we had this discussion about why returning a tuple is a bad idea

Will remove.

File: pandas/_libs/tslibs/ccalendar.pyx (hunk 5 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/ccalendar.pyx]: Well in `get_week_of_year` there _are_ array lookups, so disabling these checks should be perf-relevant, shouldn't they?

File: pandas/_libs/tslibs/ccalendar.pyx (hunk 6 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/ccalendar.pyx]: >It seems you are duplicating things from np_datetime.c ?

@jorisvandenbossche think porting less than duplicating.  The np_datetime.c version is never actually used in src/datetime (will remove old version in follow-up).  As long as perf doesn't take a hit, the default should by python/cython right?

Comment (jorisvandenbossche) [reply pandas/_libs/tslibs/ccalendar.pyx]: Sorry I don't understand your explanation. Previously it was imported it from np_datetime.c, and you reimplemented it here and changed that import. That seems like duplication rather than porting?

`dayofweek` is maybe not used in np_datetime.c itself, but `is_leapyear` is, so that one cannot just be removed there.

Comment (jbrockmendel) [reply pandas/_libs/tslibs/ccalendar.pyx]: > That seems like duplication rather than porting?

You're right.  The intention is to remove the np_datetime.c version in a follow-up PR.

`is_leapyear` _will_ be duplicated because it _is_ used in np_datetime.c, as you noted.  The idea is that np_datetime.c* is a heavy dependency (in practice tends to come with `tslibs.np_datetime` and np_datetime_strings.c) which can be replaced with a much lighter (and more pythonic) `ccalendar` dependency in some cases.

\* The name overlap between np_datetime.c and tslibs.np_datetime is not my favorite.

File: pandas/_libs/tslibs/ccalendar.pyx (hunk 5 above)
Comment (jreback) [reply pandas/_libs/tslibs/ccalendar.pyx]: where are the array lookups?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/ccalendar.pyx (hunk 5 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/ccalendar.pyx]: Line 174 lookup in month_offset

**Comment by jreback** [comment]
rebase and will look again

File: pandas/_libs/tslibs/ccalendar.pyx (hunk 6 above)
Comment (jreback) [reply pandas/_libs/tslibs/ccalendar.pyx]: I would rather include extra deps than duplicate code. mark this for a followup.

**Review by jreback — APPROVED** [review]
